### PR TITLE
Add jwt based authentication

### PR DESCRIPTION
**Pull request by VRublevski**

### CODE REVIEW

File: web/web.go [hunk 1]
```diff
@@ -311,15 +340,43 @@ func (s *Server) getVideosByID(w http.ResponseWriter, r *http.Request) {
...
-	router.HandleFunc("/requests", s.getRequests).Methods("GET")
-	router.HandleFunc("/requests/{requestId}", s.getRequestByID).Methods("GET")
+	router.HandleFunc("/requests", s.requireToken(s.getRequests)).Methods("GET")
+	router.HandleFunc("/requests/{requestId}", s.requireToken(s.getRequestByID)).Methods("GET")
 
-	router.HandleFunc("/videos", s.postVideos).Methods("POST")
-	router.HandleFunc("/videos/{operation}", s.postVideosOperation).Methods("POST")
-	router.HandleFunc("/videos/{videoId}", s.getVideosByID).Methods("GET")
+	router.HandleFunc("/videos/{operation}", s.requireToken(s.postVideosOperation)).Methods("POST")
+	router.HandleFunc("/videos/{videoId}", s.requireToken(s.getVideosByID)).Methods("GET")
```
Comment (papahmsolo) [inline web/web.go:381]: We can apply auth middleware to protected routes only using a subrouter

File: web/web.go [hunk 2]
```diff
@@ -117,7 +139,14 @@ func (s *Server) userSignUp(w http.ResponseWriter, r *http.Request) {
...
 
 func (s *Server) getRequests(w http.ResponseWriter, r *http.Request) {
-	reports, err := s.storage.ReadConversionReports("mike")
+	var login string
+	if str, ok := r.Context().Value("sub").(string); ok {
+		login = str
+	} else {
+		respondErr(w, http.StatusInternalServerError, fmt.Errorf("unable to read the login"))
+		return
+	}
```
Comment (papahmsolo) [inline web/web.go:148]: Look into `contextKey` type and methods to set and get it

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good!

File: web/web.go (hunk 1 above)
Comment (VRublevski) [reply web/web.go]: @papahmsolo I use subrouters now.

File: web/web.go (hunk 2 above)
Comment (VRublevski) [reply web/web.go]: @papahmsolo I've added a type for the login key.

File: web/web.go [hunk 3]
```diff
@@ -71,14 +80,26 @@ func (s *Server) userLogin(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		respondErr(w, http.StatusUnauthorized, err)
+		return
 	}
-	respond(w, http.StatusOK, &tkn)
+	// Set some claims
+	claims := make(jwt.MapClaims)
+	claims["sub"] = usr.Login
+	claims["exp"] = time.Now().Add(time.Hour * 72).Unix()
+	token.Claims = claims
```
Comment (papahmsolo) [inline web/web.go:95]: sc := jwt.StandardClaims{
		Subject: usr.Login,
		ExpiresAt: time.Now().Add(time.Hour * 72).Unix(),
	}

File: web/web.go [hunk 4]
```diff
@@ -117,7 +138,13 @@ func (s *Server) userSignUp(w http.ResponseWriter, r *http.Request) {
 }
 
 func (s *Server) getRequests(w http.ResponseWriter, r *http.Request) {
-	reports, err := s.storage.ReadConversionReports("mike")
+	var login string
+	var ok bool
+	if login, ok = LoginFromContext(r.Context()); !ok {
+		respondErr(w, http.StatusInternalServerError, fmt.Errorf("unable to read the login"))
+		return
+	}
```
Comment (papahmsolo) [inline web/web.go:146]:
```
	login, ok := LoginFromContext(r.Context())
	if !ok {
		respondErr(w, http.StatusInternalServerError, fmt.Errorf("unable to read the login"))
		return
	}
```

File: web/web.go [hunk 5]
```diff
@@ -311,15 +337,46 @@ func (s *Server) getVideosByID(w http.ResponseWriter, r *http.Request) {
...
+
 // Register registers handlers for endpoints in router.
 func (s *Server) Register(router *mux.Router) {
 	router.HandleFunc("/login", s.userLogin).Methods("POST")
 	router.HandleFunc("/signup", s.userSignUp).Methods("POST")
 
-	router.HandleFunc("/requests", s.getRequests).Methods("GET")
-	router.HandleFunc("/requests/{requestId}", s.getRequestByID).Methods("GET")
+	authSubRouter := router.NewRoute().Subrouter()
+	authSubRouter.Use(s.requireToken)
```
Comment (papahmsolo) [inline web/web.go:376]: Neat!

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!

File: web/web.go (hunk 3 above)
Comment (VRublevski) [reply web/web.go:95]: @papahmsolo I've changed that. Now it does not depend on the specific name of a claim.

File: web/web.go (hunk 4 above)
Comment (VRublevski) [reply web/web.go:146]: @papahmsolo Now it's less cluttered than before.

File: web/web.go (hunk 5 above)
Comment (VRublevski) [reply web/web.go:376]: @papahmsolo yep :)

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**